runelite-client/src/main/java/net/runelite/client/plugins/puzzlesolver/solver/pathfinding/IDAStarMM.java

```diff
@@ -0,0 +1,781 @@
+
+			//Check if already placed correct
+			if (locVal.getX() == x && locVal.getY() == y)
+			{
+				return;
+			}
+
+			if (locVal.getX() == locBlank.getX())
+			{
+				int diff = locBlank.getY() - locVal.getY();
+				if (diff == 1)
+				{
+					//Below
+					Point loc1 = new Point(locBlank.getX() - 1, locBlank.getY());
+					Point loc2 = new Point(loc1.getX(), loc1.getY() - 1);
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, locVal);
+				}
+				else if (diff == -1)
+				{
+					//Above
+
+					//Don't remove correct pices from row
+					if (locVal.getX() - 1 == x)
+					{
+						//Swap blank to right and continue
+						Point loc1 = new Point(locBlank.getX() + 1, locBlank.getY());
+						Point loc2 = new Point(loc1.getX(), loc1.getY() + 1);
+
+						swap(locBlank, loc1);
+						swap(loc1, loc2);
+
+						continue;
+					}
+
+					Point loc1 = new Point(locBlank.getX() - 1, locBlank.getY());
+					Point loc2 = new Point(loc1.getX(), loc1.getY() + 1);
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, locVal);
+				}
+			}
+			else if (locVal.getY() == locBlank.getY())
+			{
+				int diff = locBlank.getX() - locVal.getX();
+				if (diff == 1)
+				{
+					//Right
+
+					//Last piece
+					if (y == DIMENSION - 1)
+					{
+						Point loc1 = new Point(locVal.getX() - 1, locVal.getY() - 1);
+						Point loc2 = new Point(locVal.getX() - 1, locVal.getY());
+						Point loc3 = new Point(locVal.getX(), locVal.getY() - 1);
+						Point loc4 = new Point(locVal.getX() + 1, locVal.getY() - 1);
+
+						swap(locBlank, locVal);
+
+						swap(locVal, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+						swap(locVal, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+						swap(locVal, locBlank);
+						swap(locBlank, loc4);
+						swap(loc4, loc3);
+						swap(loc3, loc1);
+						swap(loc1, loc2);
+						swap(loc2, locVal);
+
+						return;
+					}
+
+					Point loc1 = new Point(locVal.getX() + 1, locVal.getY() + 1);
+					Point loc2 = new Point(locVal.getX(), locVal.getY() + 1);
+					Point loc3 = new Point(locVal.getX() - 1, locVal.getY() + 1);
+					Point loc4 = new Point(locVal.getX() - 1, locVal.getY());
+
+					swap(locBlank, loc1);
+					swap(loc1, loc2);
+					swap(loc2, loc3);
+					swap(loc3, loc4);
+					swap(loc4, locVal);
+				}
+				else if (diff == -1)
+				{
+					//Left
+					swap(locBlank, locVal);
+				}
+			}
+		}
+	}
+
+	private void swap(Point p1, Point p2)
+	{
+		PuzzleState newState = currentState.swap(p1.getX(), p1.getY(), p2.getX(), p2.getY());
+
+		currentState = newState;
+		stateList.add(newState);
+	}
+
+	private Point findPiece(int val)
+	{
+		for (int x = 0; x < DIMENSION; x++)
+		{
+			for (int y = 0; y < DIMENSION; y++)
+			{
+				if (currentState.getPiece(x, y) == val)
+				{
+					return new Point(x, y);
+				}
+			}
+		}
+		throw new IllegalStateException("Empty piece wasn't found!");
```

Comment:
``val`` isnt always the empty piece? Also probably add a comment "This should never happen"

Reply:
Derp, yeah ofc.